### ISSUE TITLE
FirestorePagingAdapter: using getItem(position) inside onLoadingStateChanged() loads the next page item, without any scroll

### ISSUE THREAD

**Issue by hkchakladar** [firebase/FirebaseUI-Android#1793]
I'm listening to `LOADED` in `onLoadingStateChanged` and fetching some other items based on the `documentId` of loaded documents. Here calling `getItem(int)` {last item} loads the next page items, without any scroll.

Code looks like this:

```java
...
String uid;

int mItemsCount = 0;
int mPrevLoadedCount = 0;

HashMap<Integer, Boolean> mLikesList;
HashMap<Integer, Boolean> mPreloadLikedList;
...

PagedList.Config config = new PagedList.Config.Builder()
        .setEnablePlaceholders(false)
        .setInitialLoadSizeHint(10)
        .setPrefetchDistance(5)
        .setPageSize(10)
        .build();
...

protected void onLoadingStateChanged(@NonNull LoadingState state) {
    switch (state) {
        case LOADING_INITIAL:
            mItemsCount = 0;
            mPrevLoadedCount = 0;
            Log.e("onLoadingStateChanged", "LOADING_INITIAL items: " + getItemCount());
            break;
        case LOADING_MORE:
            Log.e("onLoadingStateChanged", "LOADING_MORE items: " + getItemCount());
            break;
        case LOADED:
            ploading.setVisibility(View.GONE);
            greet.setVisibility(View.GONE);

            mItemsCount = getItemCount();

            int mmPrevLoadedCount = mPrevLoadedCount;
            int mmItemsCount = mItemsCount;

            if (uid != null) {

                while (mmPrevLoadedCount < mmItemsCount) {

//HERE is the problem. When getItem(11) {last item} is called, next page call is started, without any scroll. And this is continued till finished
                    final String memeId = getItem(mmPrevLoadedCount).getId();
                    Log.e("onLoaded", "getItem: " + mmPrevLoadedCount);


                    if (mLikesList.get(mmPrevLoadedCount) == null) {

                        final int finalMmPrevLoadedCount = mmPrevLoadedCount;

                        firestore.document("likes/" + uid + "_" + memeId).get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                            @Override
                            public void onSuccess(DocumentSnapshot documentSnapshot) {

                                if (mLikesList.get(finalMmPrevLoadedCount) == null) {

                                    boolean itemLiked = documentSnapshot.exists();
                                    mLikesList.put(finalMmPrevLoadedCount, itemLiked);

                                }
                            }
                        });
                    }
                    mmPrevLoadedCount++;
                }
            }

            mPrevLoadedCount = mItemsCount;

            Log.e("onLoadingStateChanged", "LOADED items: " + getItemCount());
            break;
        case FINISHED:
            Log.e("onLoadingStateChanged", "FINISHED items: " + getItemCount());
            break;
        case ERROR: //TODO: error home
            Log.e("onLoadingStateChanged", "ERROR items: " + getItemCount());
            //showToast("An error occurred.");
            break;
    }
}
```

Logcat:

```
2020-06-14 23:37:09.270 9066-9066/com.app E/onLoadingStateChanged: LOADING_INITIAL items: 0
2020-06-14 23:37:20.390 9066-9066/com.app E/onLoaded: getItem: 0
2020-06-14 23:37:20.392 9066-9066/com.app E/onLoaded: getItem: 1
2020-06-14 23:37:20.393 9066-9066/com.app E/onLoaded: getItem: 2
2020-06-14 23:37:20.394 9066-9066/com.app E/onLoaded: getItem: 3
2020-06-14 23:37:20.395 9066-9066/com.app E/onLoaded: getItem: 4
2020-06-14 23:37:20.396 9066-9066/com.app E/onLoaded: getItem: 5
2020-06-14 23:37:20.398 9066-9066/com.app E/onLoaded: getItem: 6
2020-06-14 23:37:20.399 9066-9066/com.app E/onLoaded: getItem: 7
2020-06-14 23:37:20.400 9066-9066/com.app E/onLoaded: getItem: 8
2020-06-14 23:37:20.401 9066-9066/com.app E/onLoaded: getItem: 9
2020-06-14 23:37:20.401 9066-9066/com.app E/onLoaded: getItem: 10
2020-06-14 23:37:20.403 9066-9066/com.app E/onLoaded: getItem: 11
2020-06-14 23:37:20.404 9066-9066/com.app E/onLoadingStateChanged: LOADED items: 12
2020-06-14 23:37:21.422 9066-9066/com.app E/onLoadingStateChanged: LOADING_MORE items: 12
2020-06-14 23:37:23.711 9066-9066/com.app E/onLoaded: getItem: 12
2020-06-14 23:37:23.712 9066-9066/com.app E/onLoaded: getItem: 13
2020-06-14 23:37:23.713 9066-9066/com.app E/onLoaded: getItem: 14
2020-06-14 23:37:23.714 9066-9066/com.app E/onLoaded: getItem: 15
2020-06-14 23:37:23.715 9066-9066/com.app E/onLoaded: getItem: 16
2020-06-14 23:37:23.716 9066-9066/com.app E/onLoaded: getItem: 17
2020-06-14 23:37:23.718 9066-9066/com.app E/onLoaded: getItem: 18
2020-06-14 23:37:23.718 9066-9066/com.app E/onLoaded: getItem: 19
2020-06-14 23:37:23.719 9066-9066/com.app E/onLoaded: getItem: 20
2020-06-14 23:37:23.720 9066-9066/com.app E/onLoaded: getItem: 21
2020-06-14 23:37:23.721 9066-9066/com.app E/onLoaded: getItem: 22
2020-06-14 23:37:23.726 9066-9066/com.app E/onLoaded: getItem: 23
2020-06-14 23:37:23.726 9066-9066/com.app E/onLoadingStateChanged: LOADED items: 24
2020-06-14 23:37:23.792 9066-9066/com.app E/onLoadingStateChanged: LOADING_MORE items: 24
2020-06-14 23:37:25.773 9066-9066/com.app E/onLoaded: getItem: 24
...
```

Remving the `getItem(int)` call, it behave normally, i.e. it loads the next page, after scrolling and reaching to end.

**Comment by samtstern** [comment]
@hkchakladar I don't quite understand what the bug in FirebaseUI is here.  The Paging library loads slightly before it is needed.  I think by calling `getItem()` you're sending that signal to the library.

**Comment by hkchakladar** [comment]
Ya, that's my question, why by calling `getItem()` it loads the next page items?

My understanding is that, it load next page items based on scrolling, when we scroll to near the last item in the recycler view. [5th item from the last, in this case].

[PagedListAdapter getItem(int)](https://developer.android.com/reference/androidx/paging/PagedListAdapter#getItem(int)) documentation also doesn't say calling `getItem()` send signal to the library to load page items.

Basically, my use case is, when the documents are loaded, like first 10 documents, I would like to get those document IDs and prefetch other documents based on the those IDs, but before showing them to the recyclerview `onBindViewHolder`.

So is there any other methods available?

**Comment by samtstern** [comment]
@hkchakladar I looked into the source code for `PagedListAdapter`.  Calling `getItem()` eventually hits this method, `loadAround()`:
https://developer.android.com/reference/androidx/paging/PagedList#loadAround(int)

The JavaDoc there says that items around the index will be loaded.  If you're at the end of the list, that means loading a new page.

So this is the intended behavior of the AndroidX library, nothing I can do within FirebaseUI unfortunately.

**Comment by hkchakladar** [comment]
OK.

As said above, my use case is, when the documents are loaded, like first 10 documents, I would like to get those document IDs and prefetch other documents based on the those IDs, but before showing them to the recyclerview onBindViewHolder.

So is there any other methods available?

**Comment by samtstern** [comment]
@hkchakladar you might be able to do what you need by calling `.getData()` on the `options` object you passed in to the list, which is a `LiveData<PagedList<DocumentSnapshot>>`.  Then you could look at the values of the `PagedList` without calling `loadAround()`.
